### PR TITLE
Set up partial functions ratchet

### PR DESCRIPTION
**Pull request by michaelpj**
- Move the `ghcide` `.hlint.yaml` to the top-level
- Add checks for a number of common partial functions, add exceptions for all existing locations
- Remove cruft from `.hlint.yaml`
- Set the hlint job to fail if there are any errors, ensuring that more instances can't be added without adding exceptions in `.hlint.yaml`.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2974"></a>

### CODE REVIEW

File: .hlint.yaml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+- warn: {name: Use module export list}
+
+# Condemn nub and friends
+- warn: {lhs: nub (sort x), rhs: Data.List.Extra.nubSort x}
+- warn: {lhs: nub, rhs: Data.List.Extra.nubOrd}
+- warn: {lhs: nubBy, rhs: Data.List.Extra.nubOrdBy}
+- warn: {lhs: Data.List.Extra.nubOn, rhs: Data.List.Extra.nubOrdOn}
+
+- functions:
+  # Things that are unsafe in Haskell base library
```
Comment (hasufell) [inline .hlint.yaml:52]: what about `unsafePerformIO`?

File: .hlint.yaml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+  # Things that are a bit dangerous in the GHC API
+  - name: nameModule
+    within:
+    - Ide.Plugin.CallHierarchy.Internal
+    - Ide.Plugin.Rename
+    - Compat.HieBin
+
+  # Partial functions
+  - name: Data.List.head
```
Comment (hasufell) [inline .hlint.yaml:71]: and `last`, `init`?

**Comment by hasufell** [comment]
CI error:

```
  Error: hlint	./ghcide/src/Development/IDE/GHC/CoreFile.hs		200	33	error	Avoid restricted function	Error in isGhcideName in module Development.IDE.GHC.CoreFile: Avoid restricted function ▫︎ Found: "nameModule" ▫︎ Note: may break the code
```

**Comment by michaelpj** [comment]
Rebased and added exceptions for the new stuff, added `unsafePerformIO` and `fold1` variants.

**Comment by michaelpj** [comment]
I think it must be picking up the hlint config.

**Comment by michaelpj** [comment]
Yep, that was it.

**Comment by michaelpj** [comment]
How can I be getting ghcide test failures?????

**Comment by hasufell** [comment]
> How can I be getting ghcide test failures?????

There can be many reasons on windows:

1. flaky tests, because windows file locking is odd (we get spurious failures in GHC/cabal/ghcup there as well)
2. github actions environments are in fact rolling-release... so a job that succeeded yesterday may now fail, e.g. due to updated msys2 or other stuff

**Comment by michaelpj** [comment]
I rebased, let's see if it reoccurs.

**Review by July541 — APPROVED** [review]

**Comment by michaelpj** [comment]
I'll leave it a day or two in case anyone else has opinions :)

**Comment by jhrcek** [comment]
Love it. Now I really don't have any excuse for not starting to remove individual partial functions one by one :smile:

**Review by eddiemundo — APPROVED** [review]

**Review by pepeiborra — APPROVED** [review]

**Pull request merged**